[TOC]

<!-- ```{css include=FALSE} -->
<!--  pre code,pre,code { -->
<!--  white-space:pre!important; -->
<!--  overflow-x: scroll!important;  -->
<!-- } -->


<!-- ``` -->


<!-- ```{r setup, include=FALSE} -->
<!-- knitr::opts_chunk$set(echo = TRUE, -->
<!--                       fig.width = 7, -->
<!--                       fig.height = 5, -->
<!--                       fig.align = "center", -->
<!--                       warning = FALSE, -->
<!--                       message = FALSE -->
<!--                       ) -->
<!-- ``` -->



## R包安装
## 数据导入
```{R eval=FALSE, include=FALSE}

# 数据包下载
# install.packages("devtools")
# library(devtools)
# devtools::install_github("ChiLiubio/meconetcomp")
# packages <- c("meconetcomp", "networkD3","ggraph","rgexf", "pheatmap", "aplot", "agricolae")
# # Now check or install
# for(x in packages){
#     if(!require(x, character.only = TRUE)) {
#         install.packages(x, dependencies = TRUE)
#     }
# }

library(devtools)
library(microeco)
library(meconetcomp)
# use pipe operator in magrittr package
library(magrittr)
library(igraph)
library(ggplot2)
# load soil amplicon sequencing dataset
library(phyloseq)
library(ape)


## 导入自己数据：
sample_info_16S = read.delim("C:/16S/result/clusternetanalysis/network_protocol-main/metadata_raw.txt",row.names = 1)
otu_table_16S = read.delim("C:/16S/result/clusternetanalysis/network_protocol-main/otutab.txt", row.names=1)
taxonomy_table_16S = read.table("C:/16S/result/clusternetanalysis/network_protocol-main/taxonomy.txt", row.names=1,header = T)
taxonomy_table_16S %<>% tidy_taxonomy
phylo_tree_16S = read.tree(file="C:/16S/result/clusternetanalysis/network_protocol-main/otus.contree")
# # tree  = read_tree("./otus.tree")
# # rep = readDNAStringSet("./otus.fa")

# 检查数据
sample_info_16S
otu_table_16S
taxonomy_table_16S
class(sample_info_16S)
class(otu_table_16S)
class(taxonomy_table_16S)
class(phylo_tree_16S)
class(dataset)

# 创建包含所有信息的microtable对象
dataset <- microtable$new(sample_table = sample_info_16S, otu_table = otu_table_16S, tax_table = taxonomy_table_16S, phylo_tree = phylo_tree_16S)
dataset

```

## 重建三个相关网络DSpre，DSmed，DSpost.
```{R eval=FALSE, include=FALSE}
# 创造列表
ora_amp_network <- list()

# 选择分组
tmp <- clone(dataset)
tmp
# 直接更改sample_table
tmp$sample_table %<>% subset(Group_dayE == "DSmed")
# 修剪对象中的所有文件
tmp$tidy_dataset()
# 使用 filter_thres 参数过滤相对丰度较低的特征
tmp <- trans_network$new(dataset = tmp, cor_method = "spearman", filter_thres = 0.00003)
tmp
# COR_p_thres 表示 p 值阈值
# COR_cut 表示相关系数阈值
tmp$cal_network(COR_p_thres = 0.05, COR_cut = 0.6)
# put the network into the list
ora_amp_network$"DSmed" <- tmp


tmp <- clone(dataset)
tmp
tmp$sample_table %<>% subset(Group_dayE == "DSpre")
tmp$tidy_dataset()
tmp <- trans_network$new(dataset = tmp, cor_method = "spearman", filter_thres = 0.00003)
tmp
tmp$cal_network(COR_p_thres = 0.05, COR_cut = 0.6)
ora_amp_network$"DSpre" <- tmp


tmp <- clone(dataset)
tmp
tmp$sample_table %<>% subset(Group_dayE == "DSpost")
tmp$tidy_dataset()
tmp <- trans_network$new(dataset = tmp, cor_method = "spearman", filter_thres = 0.00003)
tmp
tmp$cal_network(COR_p_thres = 0.05, COR_cut = 0.6)
ora_amp_network$"DSpost" <- tmp

```



## 所有网络的网络模块化；所有网络的网络拓扑属性；所有网络的节点和边属性提取.
```{R eval=FALSE, include=FALSE}
# 网络模块化
ora_amp_network %<>% cal_module(undirected_method = "cluster_fast_greedy")

# 所有网络的网络拓扑属性
tmp <- cal_network_attr(ora_amp_network)
# tmp 是一个 data.frame 对象
list(tmp)

write.table(tmp,file = "C:/16S/result/clusternetanalysis/network_protocol-main/topological_ora_DS_pmp_new.txt",quote = F,sep = '\t', row.names = T, col.names = T)

# Node and edge properties extraction for all networks 节点属性和边属性
ora_amp_network %<>% get_node_table(node_roles = TRUE) %>% get_edge_table


```


Compare nodes across networks;
Compare edges across networks;
Extract overlapped edges of networks to a new network;
Compare phylogenetic distances of paired nodes in edges

## 跨网络比较节点；
## 跨网络比较边；
## 将网络重叠边提取到新网络；
## 比较边中配对节点的系统发育距离
```{R eval=FALSE, include=FALSE}
## 绘制venn图，节点
# 通过搜索对象中的 res_node_table 获取节点分布
tmp <- node_comp(ora_amp_network, property = "name")
# 获取节点交点
tmp1 <- trans_venn$new(tmp, ratio = "numratio")
g1 <- tmp1$plot_venn(fill_color = FALSE)
g1
ggsave("C:/16S/result/clusternetanalysis/network_protocol-main/ora_amp_node_overlap_DS_pmp_new.pdf", g1, width = 7, height = 6)
# 计算 jaccard 距离来反映网络整体的差异
tmp$cal_betadiv(method = "jaccard")
tmp$beta_diversity$jaccard
node_diversity = tmp$beta_diversity$jaccard
write.table(node_diversity,file = "C:/16S/result/clusternetanalysis/network_protocol-main/ora_amp_node_diversity_DS_pmp_new.txt",quote = F,sep = '\t', row.names = T, col.names = T)


## 绘制venn图，边
# 获取跨网络的边缘分布
tmp <- edge_comp(ora_amp_network)
# 获取边的交点
tmp1 <- trans_venn$new(tmp, ratio = "numratio")
g1 <- tmp1$plot_venn(fill_color = FALSE)
g1
ggsave("C:/16S/result/clusternetanalysis/network_protocol-main/ora_amp_edge_overlap_DS_pmp_new.pdf", g1, width = 7, height = 6)
# calculate jaccard distance
tmp$cal_betadiv(method = "jaccard")
tmp$beta_diversity$jaccard
edge_diversity = tmp$beta_diversity$jaccard
# write.table(ora_amp_network_edgetax,file = "./ora_amp_edge_tax_comp.txt",quote = F,sep = '\t', row.names = T, col.names = T)
write.table(edge_diversity,file = "C:/16S/result/clusternetanalysis/network_protocol-main/ora_amp_edge_diversity_DS_pmp_new.txt",quote = F,sep = '\t', row.names = T, col.names = T)


## 比较边中成对节点的系统发育距离
## 比较跨网络边的节点源
ora_amp_network_edgetax <- edge_tax_comp(ora_amp_network, taxrank = "Phylum", label = "-", rel = TRUE)
# 过滤数值较小的特征
ora_amp_network_edgetax <- ora_amp_network_edgetax[apply(ora_amp_network_edgetax, 1, mean) > 0.02, ]
# 可视化
g1 <- pheatmap::pheatmap(ora_amp_network_edgetax, display_numbers = TRUE)
ora_amp_network_edgetax
ggsave("C:/16S/result/clusternetanalysis/network_protocol-main/ora_amp_edge_tax_comp_negtive_Phylum_DS_pmp_new.pdf", g1, width = 20, height = 7)
write.table(ora_amp_network_edgetax,file = "C:/16S/result/clusternetanalysis/network_protocol-main/ora_amp_edge_tax_comp_negtive_Phylum_DS_pmp_new.txt",quote = F,sep = '\t', row.names = T, col.names = T)


## 网络稳定性
## Robustness of network
tmp <- robustness$new(ora_amp_network, remove_strategy = c("edge_rand", "edge_strong", "node_degree_high"), 
    remove_ratio = seq(0, 0.99, 0.1), measure = c("Eff", "Eigen", "Pcr"), run = 10)
# View(tmp$res_table)
tmp$plot(linewidth = 1)
p1 <- tmp$plot(linewidth = 1)
ggsave("C:/16S/result/clusternetanalysis/network_protocol-main/ora_amp_robustness_DS_pmp_new.pdf", p1, width = 15, height = 9)


## Vulnerability of nodes.
vul_table <- vulnerability(ora_amp_network)
View(vul_table)

```




